Tickets/dm 31334

pcortes80 opened the pull request:
Please help to review this PR:

teweitsai commented on Jenkinsfile.labview:
```diff
@@ -36,7 +36,7 @@ pipeline {
                     withEnv(["HOME=${env.WORKSPACE}"]) {
                         sh """
-                            xdotool key --repeat 5 --delay 1000 Return | timeout 60 labview64 tests/testAllWithXmlReport.vi || true
+                            xdotool key --repeat 5 --delay 1000 Return | timeout 60 labview64 tests/testAllWithXmlReport.vi || true                            		    
```
Please remove all white trails (spaces + tabs) after the `|| true`:
![image](https://user-images.githubusercontent.com/11474902/132734689-7158bd27-6b29-4a57-b395-45ac1e5cbe64.png)

pcortes80 replied:
Done.

teweitsai commented on doc/versionHistory.md:
```diff
@@ -1,5 +1,10 @@
 # Version History
 
+0.2.2
+
+- TCP Server can receive telemetry from TCP Client.
+- TCP Server can send telemetry to TCP Client.
+
```
Please also say this version moves the attributes of `firstPacket` and `lastSequenceId` as from **tcpServerBase** to **tcpServerCmd** with great details. Thanks!

pcortes80 replied:
Done.

teweitsai reviewed (changes requested):
Rewrite.

pcortes80 commented on simulator/README.md:
```diff
@@ -83,10 +74,45 @@ Note:
+Note:
+
+1. Before publish this telemetry you have to subscribe to this telemetry in TCP Client.
+
```
Done.

pcortes80 commented on simulator/README.md:
```diff
@@ -31,15 +31,6 @@ From LabVIEW project:
 
 Go to the `simulator/` directory, find `ComponentSimulator.vi`, press right click on it and select **Run**.
 
-Check:
-
-- cmdServerPort = 8888
-- telServerPort = 8889
-- serverTimeout (ms) = 100
-- serverBytesToRead = 32768
-
-Note: this configuration cannot be modified in runtime.
-
```
lines are back. Sorry.

pcortes80 commented on simulator/README.md:
```diff
@@ -83,10 +74,45 @@ Note:
+
+1. Before publish this telemetry you have to subscribe to this telemetry in TCP Client.
+
+2. Is recommended add more time in Duration option when subscribe to this telemetry in TCP Client.
```
Done. Thanks.

pcortes80 commented on simulator/README.md:
```diff
@@ -83,10 +74,45 @@ Note:
 
+### Receive Telemetry from TCP Client
+
+- Press "Get Queue Size" buttom to know how many elements has the queue.
+
+- Press "Flush Telemetry" buttom to remove all elements in queue. They will be showed in "Telemetry Flushed" indicator.
+
+- Press "Get on instance Telemetry" to read telemetry in queue one by one. You can see the elements is queue in "Remaining Elements" indicator.
+
```
Done. Thanks.

pcortes80 commented on src/README.md:
```diff
@@ -83,3 +83,19 @@ Note:
+### TcpServer design details
+
+- `tcpServerSenderLoop.vi` is set as preallocated clone reentrant execution, in **tcpServerBase**, **tcpServerCmd** and **tcpServerTel** classes.
+
```
Done. Thanks.

pcortes80 commented on src/README.md:
```diff
@@ -83,3 +83,19 @@ Note:
+- `tcpServerSenderLoop.vi` is set as preallocated clone reentrant execution, in **tcpServerBase**, **tcpServerCmd** and **tcpServerTel** classes.
+
+- `RunServer.vi` works as Dynamic Dispatch in **tcpServerBase**, **tcpServerCmd** and **tcpServerTel** classes.
+
```
Done. Thanks.

pcortes80 commented on src/README.md:
```diff
@@ -83,3 +83,19 @@ Note:
+- `RunServer.vi` works as Dynamic Dispatch in **tcpServerBase**, **tcpServerCmd** and **tcpServerTel** classes.
+
+- `tcpServerStateMachine.vi` works as Dynamic Dispatch in **tcpServerBase**, **tcpServerCmd** and **tcpServerTel** classes.
+
```
Done. Thanks.

pcortes80 commented on src/README.md:
```diff
@@ -83,3 +83,19 @@ Note:
+- `tcpServerStateMachine.vi` works as Dynamic Dispatch in **tcpServerBase**, **tcpServerCmd** and **tcpServerTel** classes.
+
+- User can configure TCP Server to work with a circular buffer when sends telemetry to TCP Client and receives telemetry from TCP Client, through `configServer.vi` when server starts. In **queueOptions** control select **lossy enqueue** to use circula buffer, and **normal enqueue** to use standar FIFO technique.
+
```
Done. Thanks.

pcortes80 commented on src/README.md:
```diff
@@ -83,3 +83,19 @@ Note:
+- User can configure TCP Server to work with a circular buffer when sends telemetry to TCP Client and receives telemetry from TCP Client, through `configServer.vi` when server starts. In **queueOptions** control select **lossy enqueue** to use circula buffer, and **normal enqueue** to use standar FIFO technique.
+
+- User can enable/disable circular buffer through `sendTel.vi` in runtime, only to send telemetry to TCP Client.
+
```
Done. Thanks.

pcortes80 commented on src/README.md:
```diff
@@ -83,3 +83,19 @@ Note:
+- User can enable/disable circular buffer through `sendTel.vi` in runtime, only to send telemetry to TCP Client.
+
+- User can configure TCP Server as **block call** code when sends telemetry to TCP Client through `sendTel.vi` with Timeout = -1 and queueOption = normal enqueue. To set this as **un-block call** use Timeout greater than 0 and queueOptions = lossy enqueue.
+
```
Done. Thanks.

pcortes80 commented on src/README.md:
```diff
@@ -83,3 +83,19 @@ Note:
+
+- `tcpServerSenderLoop.vi` is set as preallocated clone reentrant execution, in **tcpServerBase**, **tcpServerCmd** and **tcpServerTel** classes.
+
+- `RunServer.vi` works as Dynamic Dispatch in **tcpServerBase**, **tcpServerCmd** and **tcpServerTel** classes.
```
Picked  `runServer.vi` . Thanks.

pcortes80 commented on src/README.md:
```diff
@@ -83,3 +83,19 @@ Note:
+
+- `tcpServerStateMachine.vi` works as Dynamic Dispatch in **tcpServerBase**, **tcpServerCmd** and **tcpServerTel** classes.
+
+- User can configure TCP Server to work with a circular buffer when sends telemetry to TCP Client and receives telemetry from TCP Client, through `configServer.vi` when server starts. In **queueOptions** control select **lossy enqueue** to use circula buffer, and **normal enqueue** to use standar FIFO technique.
```
Fixed. Thanks.

pcortes80 commented on src/README.md:
```diff
@@ -83,3 +83,19 @@ Note:
+
+- User can enable/disable circular buffer through `sendTel.vi` in runtime, only to send telemetry to TCP Client.
+
+- User can configure TCP Server as **block call** code when sends telemetry to TCP Client through `sendTel.vi` with Timeout = -1 and queueOption = normal enqueue. To set this as **un-block call** use Timeout greater than 0 and queueOptions = lossy enqueue.
```
Done. Thanks.

teweitsai commented on simulator/README.md:
```diff
@@ -31,12 +31,16 @@ From LabVIEW project:
 - serverBytesToRead = 32768
+- maxSizeQueueTel = 1000
+- EnqueueOptions = lossy enqueue
+- timeoutQueueTel = 50
```
Maybe `timeoutQueueTel (ms) = 50` if I am right for the unit. Thanks!

pcortes80 replied:
Done. Thanks.

teweitsai commented on simulator/README.md:
```diff
@@ -83,10 +87,44 @@ Note:
+
+| Telemetry Id |   Telemetry Details    |
+|:------------:|:----------------------:|
+|   elevation  | "compName" : "MTMount" |
```
In this case, maybe:
|   elevation  | "measuredInDeg" : 30 |

This is because this telemetry is published from the LabVIEW component. Therefore, you should not see the field of "compName", which happens in the subscription part.

pcortes80 replied:
Good to know. Thanks!

teweitsai commented on simulator/README.md:
```diff
@@ -83,10 +87,44 @@ Note:
+Note:
+
+1. Before publish this telemetry you have to subscribe to this telemetry in TCP Client.
+2. It is recommended add more time in Duration option when subscribe to this telemetry in TCP Client.
```
It is recommended add more time ... --> It is recommended to add more time ...

pcortes80 replied:
Done. Thanks!

teweitsai commented on src/README.md:
```diff
@@ -83,3 +84,16 @@ Note:
 1. The polymorphic `convertToKeyValuePair.vi` does not support `Inf` value.
 Please try to avoid use this value.
+
+### TcpServer design details
```
TcpServer design details --> TcpServer Design Details

pcortes80 replied:
of course capital letter! Thanks!

teweitsai reviewed (approved):
[ComponentSimulator.vi]

1. In the event case of "publish_telemetry", the TcpServer.sendTel.vi should read the input of queueOptions from the front panel as well (you did this in the event case of "send_telemetry".

**Test**

[testIsTelQueueFull.vi]

1. I think you can call the TcpServer.sendTel.vi to write the telemetry elements to simplify the code.
2. I think you can call the TcpServer.flushTelQueue.vi to flush the elements to simplify the code.

**TcpServer**

[getTelElement.vi]

1. I think you should have a TcpServerTel.getTelElement.vi and let the TcpServer.getTelElement.vi to call it directly.

[flushTelQueue.vi]

1. Are you sure we are flushing the queue with telemetry that come from TCP Client? Right now, we have two queue here: (1) Python TcpClient to the TcpServer, and (2) LabVIEW component to the TcpServer. You could see the attributes names are TcpServerBase.tcpClientQueue and TcpServerTel.queueTel. The name are confusing. In addition, how do you know which queue that the user wants to flush? It might be better to have an enum to differentiate this.

[isTelQueueFull.vi]

1. We have two queues here (see the comments in flushTelQueue.vi). How do you know which queue that the user wants to know it is full or not? It might be better to have an enum to differentiate this.

**TcpServerTel**

[flushTel.vi]

1. Same comment in TcpServer.flushTelQueue.vi.

[isTelQueueFull.vi]

1. Same comment in TcpServer.isTelQueueFull.vi.

[checkQueueFull.vi]

1. Same comment in TcpServer.isTelQueueFull.vi.

[enqueueTelToComponent.vi]

1. I think the timeout value in the configuration should be read in the "Enqueue Element" in the case of "normal enqueue"?

[processTel.vi]

1. If there is a Fault, we should pass this information in the LabVIEW component. You have a ticket for this. Right? Please put the ticket number to the label of "NOTE:".

[writeToClient.vi]

1. Please remove this vi if no one calls it.

teweitsai commented:
I only take a quick look in the refactoring part of **TcpServerCmd** class. I think we may need to work on the DM-31606 first. Let us test the code together and decide the priority of JIRA tickets. Thanks!

pcortes80 commented:
[ComponentSimulator.vi]

1. In the event case of "publish_telemetry", the TcpServer.sendTel.vi should read the input of queueOptions from the front panel as well (you did this in the event case of "send_telemetry").

Reply:
- Fixed. Thanks!

**Test**

[testIsTelQueueFull.vi]

I made a mistake in TcpServer.isTelQueueFull.vi. What I understood from you is to have a public vi in TcpServer that user can use to know if queue is full or not to send telemetry FROM COMPONENT TO CLIENT. So, this vi has a wrong name, because the queue to send telemetry to client is called tcpClientQueue. I updated the vi name to isTcpClientQueueFull.vi, so the test vi name now is testIsTcpClientQueueFull.vi

1. I think you can call the TcpServer.sendTel.vi to write the 
telemetry elements to simplify the code.

Reply:
- Good idea. Done. Thanks!

2. I think you can call the TcpServer.flushTelQueue.vi to flush the elements to simplify the code.

Reply:
- This is not possible, because this function flush the queueTel that -receives- telemetry from client, and I want to flush the queue that -sends- telemetry to client. That is why I get the queue from base and flush queue with the flush labview function.

**TcpServer**

[getTelElement.vi]

1. I think you should have a TcpServerTel.getTelElement.vi and let the TcpServer.getTelElement.vi to call it directly.

Reply:
- Done. Thanks!

[flushTelQueue.vi]

1. Are you sure we are flushing the queue with telemetry that come from TCP Client? Right now, we have two queue here: (1) Python TcpClient to the TcpServer, and (2) LabVIEW component to the TcpServer. You could see the attributes names are TcpServerBase.tcpClientQueue and TcpServerTel.queueTel. The name are confusing. In addition, how do you know which queue that the user wants to flush? It might be better to have an enum to differentiate this.

Reply:
- yes, I am. In this vi we are flushing the queue with telementry that comes from TCP Client.
- Actually we have these queues: (1) Python TcpClient to the TcpServer <tcpQueueServer>, (2) TcpServer to LabVIEW component after processing the telemetry <queueTel> and (3) LabVIEW component to the TcpServer <tcpQueueClient>. I only flush the queueTel, because this is the accepeted telemetry in TcpServer. 
- Maybe we can modify the name of every queue, because you are right they are confusing. Can we discuss this?

[isTelQueueFull.vi]

1. We have two queues here (see the comments in flushTelQueue.vi). How do you know which queue that the user wants to know it is full or not? It might be better to have an enum to differentiate this.

Reply:
- as I commented before, the only queue that user wants to know if is full or not is tcpQueueClient, the one used to send telemetry from LabVIEW component to TcpServer, as I understood from you.

**TcpServerTel**

[flushTel.vi]

1. Same comment in TcpServer.flushTelQueue.vi.

Reply:
- as I commented before, the only queue that user wants to flush is queueTel, the one used to send telemetry from TcpServer to LabVIEW component, after telemetry is proccessed by TcpServer.

[isTelQueueFull.vi]

1. Same comment in TcpServer.isTelQueueFull.vi.

Reply:
- in TcpServerTel there is no isTelQueueFull.vi. Insted there is checkQueueFull.vi. This one checks if tcpClientQueue is full or not. I just modified this. 
- as I commented before, the only queue that user wants to know if is full or not is tcpQueueClient, the one used to send telemetry from LabVIEW component to TcpServer, as I understood from you. 

[checkQueueFull.vi]

1. Same comment in TcpServer.isTelQueueFull.vi.

Reply:
- same reply in [isTelQueueFull.vi].

[enqueueTelToComponent.vi]

1. I think the timeout value in the configuration should be read in the "Enqueue Element" in the case of "normal enqueue"?

Reply:
- Fixed. Thanks.

[processTel.vi]

1. If there is a Fault, we should pass this information in the LabVIEW component. You have a ticket for this. Right? Please put the ticket number to the label of "NOTE:".

Reply:
- Done. Thanks.

[writeToClient.vi]

1. Please remove this vi if no one calls it.

Reply:
- Removed. Thanks.
